File: transient_random.py
# =============================================================================
# This script generates random 2-D fields, computes heads and
# stream function values, constructs streamline-oriented grids,
# computes transient concentrations for a step-input inflow condition, 
# and evaluates breakthrough curves for all stream tubes
#
# written by 
#
# Olaf A. Cirpka
#
# University of Tuebingen
# Department of Geosciences
# Schnarrenbergstr. 94-96
# 72076 Tuebingen
# Germany
# olaf.cirpka@uni-tuebingen.de
#
# Matlab Code: September 19, 2013; updated October 4, 2019
# transferred to python: April 2025
# =============================================================================

# clear all variables
from IPython import get_ipython
get_ipython().run_line_magic('reset', '-sf')
# clear console
import os
os.system('cls' if os.name == 'nt' else 'clear')

import numpy as np
import matplotlib.pyplot as plt
from scipy.sparse.linalg import spilu, LinearOperator, bicgstab
import pyamg # installation of pyamg: pip install pyamg
import time
from types import SimpleNamespace
from datetime import datetime

from fullscreenfigure import fullscreenfigure
from randomK import randomK
import gw_FEM as gw
from streamlinegrid import slgrid, store_mat, mob_mat, quad_cell_areas

# number of elements per direction (x,y)
nx = [500, 250]
# grid spacing
dx = [0.01, 0.01]
# correlation lengths
lx = [.5, .1]
# rotation angle
ang= 0.*np.pi/180.
# variance of ln(K)
sigY= 1.
# type of autocovariance function
Ctype = 1
# geometric mean of conductivity
Kg = 1e-4

# head difference
phiin = nx[0]*dx[0]*0.01

# transport parameters
poros = 0.3    # porosity [-]
al    = 0.01   # longitudinal dispersivity [m]
at    = 0.001  # transverse dispersivity [m]
Dm    = 1e-9   # molecular diffusion coefficient [m]
CN=1.0         # Crank-Nicolson weight for time integration
plottransient=False # flag whether transient profiles should be plotted

# dimensions of the streamline-oriented grid
ntube = 250
nsec = 500

# nodal coordinates
X = np.arange(nx[0]+1)*dx[0]
Y = np.arange(nx[1]+1)*dx[1]
(X,Y)=np.meshgrid(X,Y)
# number of nodes
nnod = (nx[0]+1)*(nx[1]+1)

# -----------------------------------------------------------------------------
# generate conductivity field
# -----------------------------------------------------------------------------
print("Generate Random Field")
K = randomK(nx, dx, lx, ang, sigY, Ctype, Kg)

# -----------------------------------------------------------------------------
# groundwater flow
# -----------------------------------------------------------------------------
tic = time.time()
print("Groundwater Flow")
# set up system of equations for groundwater flow
print("Set up System of Equations")
M = gw.mat(K,dx[0],dx[1])

# Account for fixed-head boundary conditions
print("Account for Fixed-Head Boundary Conditions")
leftnodes  = np.arange(nx[1]+1)*(nx[0]+1)
rightnodes = leftnodes + nx[0]
r = np.zeros(nnod)
Mmod, rmod = gw.diri(M,r,leftnodes,phiin*np.ones((nx[1]+1,1)))
Mmod, rmod = gw.diri(Mmod,rmod,rightnodes,np.zeros((nx[1]+1,1)))
# Preconditioner
print("Preconditioning of the System of Equations")
# ILU preconditioning (if pyamg is not available)
# ilu = spilu(Mmod)
# Mmod2 = LinearOperator(Mmod.shape, ilu.solve)
# AMG preconditioning (requires pyamg)
ml = pyamg.ruge_stuben_solver(Mmod)
Mmod2 = ml.aspreconditioner()
# solve groundwater-flow equation
h0=(1-np.reshape(X,(nnod,1))/dx[0]/nx[0])*phiin
print("Solve System of Equations")
h, info = bicgstab(Mmod, np.array(rmod)[:,None],rtol=1e-8,atol=1e-12,M=Mmod2,x0=h0)
# evaluate total flux
Qin = sum(M[leftnodes,:]@h)

# -----------------------------------------------------------------------------
# stream function
# -----------------------------------------------------------------------------
print("Stream Function")
# set up system of equations for stream-function equation
print("Set up System of Equations")
M = gw.mat(K**-1,dx[0],dx[1])

# Dirichlet boundary conditions
print("Account for Fixed-Value Boundary Conditions")
botnodes = np.arange(nx[0]+1)
topnodes = botnodes + nx[1]*(nx[0]+1)
r = np.zeros(nnod)
Mmod, rmod = gw.diri(M,r,botnodes,np.zeros((nx[0]+1,1)))
Mmod, rmod = gw.diri(Mmod,rmod,topnodes,Qin*np.ones((nx[0]+1,1)))
# Preconditioner
print("Preconditioning of the System of Equations")
# ILU preconditioning (if pyamg is not available)
# ilu = spilu(Mmod)
# Mmod2 = LinearOperator(Mmod.shape, ilu.solve)
# AMG preconditioning (requires pyamg)
ml = pyamg.ruge_stuben_solver(Mmod)
Mmod2 = ml.aspreconditioner()
print("Solve System of Equations")
psi0=(1-np.reshape(Y,(nnod,1))/dx[1]/nx[1])
psi, info = bicgstab(Mmod, np.array(rmod)[:,None],rtol=1e-8,atol=1e-16,M=Mmod2,x0=psi0)
toc = time.time()
print(f"Elapsed time for head and stream-function calculation: {toc - tic:.4f} seconds")
tic = toc

psi = np.reshape(psi,(nx[1]+1,nx[0]+1))
h   = np.reshape(h,(nx[1]+1,nx[0]+1))

# -----------------------------------------------------------------------------
# streamline-oriented grid generation
# -----------------------------------------------------------------------------
print("Streamline-Oriented Grid Generation")
net = slgrid(ntube, nsec, nx, dx, psi, h, phiin, Qin)
toc = time.time()
print(f"Elapsed time for grid generation: {toc - tic:.4f} seconds")
tic = toc

# np.savez('mydata.npz',K=K,h=h,psi=psi,Qin=Qin,xnet=net.x,ynet=net.y,
#          psinet=net.psi,phinet=net.phi)

# data = np.load('mydata.npz',allow_pickle=True)
# psi = data['psi']
# h = data['h']
# K = data['K']
# Qin = data['Qin']
# net = SimpleNamespace(
#       x=data["xnet"], y=data["ynet"],
#       phi=data["phinet"], psi=data["psinet"])

# -----------------------------------------------------------------------------
# Preparation of Transport Calculations
# -----------------------------------------------------------------------------
print("Prepare Matrices for Transport Calculations")
# water-filled area of all cells
porarea = quad_cell_areas(net.x, net.y)*poros

# calculate matrices to solve for ADE on streamline-oriented grid
Mstore = store_mat(net, poros)              # storage matrix
Mmob   = mob_mat(net, al, at, Dm, poros, 1) # mobility matrix

# -----------------------------------------------------------------------------
# Compute Mean Travel Time by Groundwater-Age Equation
# -----------------------------------------------------------------------------
print("Solve Mean Groundwater-Age Equation")
rhs=porarea.reshape((ntube*nsec,1))
# Preconditioner
ilu = spilu(Mmob)
Mmob2 = LinearOperator(Mmob.shape, ilu.solve)
# solve for first temporal moment
m1, info = bicgstab(Mmob,rhs,rtol=1e-8,atol=1e-12,M=Mmob2)
m1=m1.reshape(ntube,nsec)

toc = time.time()
print(f"Elapsed time for calculation of 1st moment: {toc - tic:.4f} seconds")

# -----------------------------------------------------------------------------
# Transient transport
# -----------------------------------------------------------------------------
# prepare transient simulation
# total time
tend  = 1.5*np.max(m1)
if CN<1:
   # determine time step size so that maximum Courant number is unity
   dt = np.min(porarea)/Qin*ntube
else:
   # determine time step size so that mean Courant number is unity
   dt = np.mean(porarea)/Qin*ntube
tvec= np.arange(dt,tend,dt)

# initialize breakthrough curve
BTC = np.zeros((ntube,len(tvec)))
lastcol = list(range(nsec-1,nsec*ntube,nsec))

# Remark: Use Crank-Nicolson for transport
# (Mstore/dt + CN*Mmob)*c_new = (Mstore/dt - (1-CN)*Mmob)*c_old + r_source
Mleft  = Mstore/dt + CN*Mmob
Mright = Mstore/dt - (1-CN)*Mmob

# initialization
c = np.zeros((ntube*nsec,1))

# specific discharge per stream tube
invec = np.zeros((ntube*nsec,1))
invec[0:nsec*ntube:nsec]=Qin/ntube

cin = np.zeros((ntube*nsec,1))
cin[0:nsec*ntube:nsec]=1.


# AMG preconditioner
ml = pyamg.air_solver(Mleft)
Mleft2 = ml.aspreconditioner()


print('Solve for Transient Concentrations')
t=0.
tdt=datetime.utcfromtimestamp(t)
tdays=0.

# initialize plot
if plottransient:
   fig2 = fullscreenfigure(2)
   ax = fig2.add_subplot()
   ax.cla()
   cplot=ax.pcolormesh(net.x, net.y, c.reshape((ntube,nsec)), 
                       shading='flat', cmap='jet', vmin=0., vmax=1.)
   plt.colorbar(cplot, label='c/c$_{in}$ [-]')
   ax.set_aspect(aspect=1.0)
   ax.set_xlabel('x [m]')
   ax.set_ylabel('y [m]')
   titext = ax.set_title(f'Transient Concentration after{tdays:3.0f} days, {tdt.strftime("%H:%M:%S")}')
   plt.draw()
   plt.pause(0.1)
   plt.show()

# time loop
for i in range(len(tvec)):
    t = (i+1)*dt
    print(f't = {t:10.0f} s')
    BTCold=c[lastcol]
    # update right-hand side vector
    rhs=Mright @ c + np.ones((ntube*nsec,1))*invec*cin
    # solve for concentration
    c, info = bicgstab(Mleft, rhs,rtol=1e-8,atol=1e-16,M=Mleft2,x0=c)
    c = c.reshape((ntube*nsec,1))
    # breakthrough curves in the outflow
    BTC[:,i]=(1-CN)*BTCold[:,0]+CN*c[lastcol,0]
    # stop loop if breakthrough is complete
    if np.min(BTC[:,i]) > 0.999:
       tvec=tvec[:i+1]
       BTC = BTC[:,:i+1]
       break
    # update plot
    if plottransient:
       tdt=datetime.utcfromtimestamp(t)
       tdays = np.floor(t/86400)
       cplot.set_array(c.reshape((ntube,nsec)))
       titext.set_text(
 f'Transient Concentration after{tdays:3.0f} days, {tdt.strftime("%H:%M:%S")}')
       # updated values
       fig2.canvas.draw()
       fig2.canvas.flush_events()
       plt.pause(0.1)
       plt.show()

# -----------------------------------------------------------------------------
# Travel-Time Distributions
# -----------------------------------------------------------------------------
# take derivative of BTCs to obtain travel-time pdf of each streamtube
ptau = np.zeros((ntube,len(tvec)))
ptaurel = np.zeros((ntube,len(tvec)))
trel=np.outer(m1[:,-1]**-1,tvec)
for ii in range(ntube):
    ptau[ii,:]=np.gradient(BTC[ii,:],tvec)
    ptaurel[ii,:]=np.gradient(BTC[ii,:],trel[ii,:]);

# -----------------------------------------------------------------------------
# REPLACEMENT SCENARIO  A+B > C with 1:1:1 stoichiometry
# Initial concentration of B: 1
# Injected concentration of A: 1
# -----------------------------------------------------------------------------
# Use local breakthrough curves
BTC_A=2*BTC-1
BTC_A[BTC_A<0] = 0
BTC_B=1-2*BTC
BTC_B[BTC_B<0] = 0
BTC_C=BTC-BTC_A

# use averaged BTC of mixing ratio
BTC_Am=2*np.mean(BTC,0)-1
BTC_Am[BTC_Am<0]=0
BTC_Bm=1-2*np.mean(BTC,0)
BTC_Bm[BTC_Bm<0]=0
BTC_Cm=np.mean(BTC,0)-BTC_Am


# plot conductivity field
fig1 = fullscreenfigure(1)
ax1  = plt.subplot(2,2,1)
pc   = ax1.pcolormesh(X, Y, np.log10(K), shading='flat', cmap='jet')
ax1.set_aspect(aspect=1.0)
plt.colorbar(pc, label='log$_{10}$K [K in m/s]')
ax1.set_xlabel('x [m]')
ax1.set_ylabel('y [m]')
ax1.set_title('Log-Conductivity Field')
# plot flownet
ax2 = plt.subplot(2,2,2)
ax2.contour(X,Y,psi,50,colors='k')
co  = ax2.contour(X,Y,h,50,cmap='jet')
ax2.set_aspect(aspect=1.0)
cb2 = plt.colorbar(co, label='h [m]')
ax2.set_xlabel('x [m]')
ax2.set_ylabel('y [m]')
ax2.set_title('Flownet')
# plot streamline-oriented grid
ax3 = plt.subplot(2,2,3)
ax3.plot(net.x,net.y,'k-', linewidth = 0.5)
ax3.plot(net.x.T,net.y.T,'k-',linewidth = 0.5)
ax3.set_aspect(aspect=1.0)
ax3.set_xlabel('x [m]')
ax3.set_ylabel('y [m]')
ax3.set_title('Streamline-Oriented Grid')
ax3.set_xlim(0,nx[0]*dx[0])
ax3.set_ylim(0,nx[1]*dx[1])
# plot mean groundwater age
ax4 = plt.subplot(2,2,4)
m1plot=ax4.pcolormesh(net.x, net.y, m1/86400, shading='flat', cmap='jet')
plt.colorbar(m1plot, label='\u03C4 [d]')
ax4.set_aspect(aspect=1.0)
ax4.set_xlabel('x [m]')
ax4.set_ylabel('y [m]')
ax4.set_title('Mean Groundwater Age')
ax4.set_xlim(0,nx[0]*dx[0])
ax4.set_ylim(0,nx[1]*dx[1])

# plot travel-time distributions
fig3 = fullscreenfigure(3)
plt.subplot(2,1,1)
plt.plot(np.outer(np.ones(ntube),tvec).T/86400,ptau.T*86400,'#808080', linewidth = 0.5)
plt.plot(tvec/86400,np.mean(ptau,0)*86400,'k',linewidth=4)
plt.xlabel('\u03C4 [d]')
plt.ylabel('p(\u03C4) [1/d]')
plt.title('Local Travel-Time Distributions at the Outlet Face')
plt.subplot(2,1,2)
plt.plot(trel.T,ptaurel.T,'#808080', linewidth = 0.5)
plt.xlabel('\u03C4$_{rel}$ [-]')
plt.ylabel('p(\u03C4$_{rel}$) [-]')
plt.title('Local Travel-Time Distributions Scaled by Local Mean Groundwater Age')

# plot mean and standard deviation of travel-time distribution
fig4 = fullscreenfigure(4)
plt.plot(tvec/86400,np.mean(ptau,0)*86400,label='mean of p(\u03C4)')
plt.plot(tvec/86400,np.std(ptau,0)*86400,label='std. dev. of p(\u03C4)')
plt.xlabel('\u03C4 [d]')
plt.ylabel('p(\u03C4) [1/d]')
plt.legend()
plt.title('Mean and Stndard Deviation of Local Travel-Time Distributions')

# plot concentrations of mixing-controlled reactive transport
fig5 = fullscreenfigure(5)
plt.subplot(2,1,1)
plt.plot(np.outer(np.ones(ntube),tvec).T/86400,BTC_A.T,'#76a5d2',linewidth=0.5)
plt.plot(np.outer(np.ones(ntube),tvec).T/86400,BTC_B.T,'#e88c8f',linewidth=0.5)
plt.plot(np.outer(np.ones(ntube),tvec).T/86400,BTC_C.T,'#7ac67a',linewidth=0.5)
plt.plot(tvec/86400,np.mean(BTC_A,0),'b',linewidth=4,label='A')
plt.plot(tvec/86400,np.mean(BTC_B,0),'r',linewidth=4,label='B')
plt.plot(tvec/86400,np.mean(BTC_C,0),'g',linewidth=4,label='C')
plt.xlabel('t [d]')
plt.ylabel('c$_{reac}$')
plt.legend()
plt.title('Reactive-Species Concentrations at the Outlet Face')

plt.subplot(2,1,2)
plt.plot(tvec/86400,np.mean(BTC_A,0),'b',label='correct average')
plt.plot(tvec/86400,np.mean(BTC_B,0),'r')
plt.plot(tvec/86400,np.mean(BTC_C,0),'g')
plt.plot(tvec/86400,BTC_Am,'b--',label='from average mixing ratio')
plt.plot(tvec/86400,BTC_Bm,'r--')
plt.plot(tvec/86400,BTC_Cm,'g--')
plt.xlabel('t [d]')
plt.ylabel('c$_{reac}$')
plt.legend()
plt.title('Mean Reactive-Species Concentrations at the Outlet Face')

plt.show(block=True)

File: transient_random.m
% MATLAB script transient_random.m
%
% This script generates random 2-D fields, computes heads and
% stream function values, constructs streamline-oriented grids,
% computes transient concentrations for a step-input inflow condition, 
% and evaluates breakthrough curves for all stream tubes
%
% written by 
%
% Olaf A. Cirpka
%
% University of Tuebingen
% Center for Applied Geoscience
% Hoelderlinstr. 12
% 72074 Tuebingen
% Germany
% Olaf.Cirpka@uni-tuebingen.de
%
% September 19, 2013
% update October 4, 2019


clear all;close all;clc
% ============== BEGIN INPUT BLOCK ================================================
% number of elements per direction
nx = [500,250];
% grid spacing
dx = [.01,.01];
% correlation length
lx = [.50, .10];
% rotation angle
ang = 0;
% Exponential (1) or Gaussian (2) Covariance-Model 
Ctype = 1;
% number of stream tubes
ntube = 250;
% number of sections per stream tube
nsec =  500;
% geometric mean of conductivity 
Kg = 1e-4;
% variance of ln(K)
varY=1;
% head difference
phiin = nx(1)*dx(1)*0.01;
% transport parameter
poros = 0.3;
al    = 0.01;
at    = 0.001;
Dp    = 1e-9;
% plot transient concentration rsults?
plottransient=true;
% Crank-Nicolson weight for time integration (0: explicit, 1: implicit)
CN=0.5;
% ============== END INPUT BLOCK ==================================================

doagain=true;
while doagain
% ============== BEGIN FIELD GENERATION BLOCK =====================================
% initialize random-number generator
rng(sum(100*clock),'twister');
% Spatial coordinates
[X,Y]=meshgrid([0:nx(1)]*dx(1),[0:nx(2)]*dx(2));
% Generate the field
K=randomK(nx,dx,lx,ang,varY,Ctype,Kg);

% Output
figure(1)
colormap('jet')
set(gcf,'outerposition',get(0,'screensize'),'paperunits','centimeters',...
    'paperposition',[0 0 20 10])
pcolor(X,Y,log10([K K(:,1);K(1,:) K(1,1)]));
shading flat;
xlim([0,nx(1)*dx(1)]);
ylim([0,nx(2)*dx(2)]);
box on;
daspect([1 1 1]);
xlabel('x_1');ylabel('x_2');
cb=colorbar;
ylabel(cb,'log_{10} K (K in m/s)')
title('log-conductivity field')
drawnow;
% print -djpeg100 -r300 Kfield.jpg
% ============== END FIELD GENERATION BLOCK =======================================

% ============== BEGIN HEAD CALCULATION ===========================================
% stiffness matrix of a single element
disp([datestr(clock) ': Set up System of Equations for Head']);
M_h_el = calc_h_stiff_el(dx(2),dx(1));
% stiffness matrix of entire system
M      = sparprep(K,M_h_el);
% number of nodes
nnod = prod(nx+1);
% righthand side vector
r = zeros(nnod,1);
% in- and outflow nodes
innod  = [1:nx(2)+1];
outnod = innod + nx(1)*(nx(2)+1);
% set Dirichlet boundary condition
[Mmod,rmod]=gwdiri(M,r,[innod outnod],[phiin*ones(1,nx(2)+1) zeros(1,nx(2)+1)]);

% LU decomposition
% disp([datestr(clock) ': Incomplete LU decomposition']);
% [L,U] = ilu(Mmod);
disp([datestr(clock) ': Solve for Head']);
% h     = bicgstab(Mmod,rmod,1e-12,10*nnod,L,U);
h     = Mmod\rmod;
% remove roundoff errors of inflow and outflow nodes
h(innod)=phiin;
h(outnod)=0;

Qin = sum(M(innod,:)*h);
disp([datestr(clock) sprintf(': Qin = %g m2/s',Qin)]);
% ============== END HEAD CALCULATION =============================================

% ============== BEGIN STREAM FUNCTION CALCULATION ================================
% stiffness matrix of entire system
disp([datestr(clock) ': Set up System of Equations for Stream Function']);
M    = sparprep(K.^-1,M_h_el);

% righthand side vector
r = zeros(nnod,1);
% in- and outflow nodes
botnod = [1:nx(2)+1:nnod];
topnod = botnod + nx(2);
% boundary condition
[Mmod,rmod]=gwdiri(M,r,[botnod topnod],[zeros(1,nx(1)+1) Qin*ones(1,nx(1)+1)]);

% LU decomposition
% disp([datestr(clock) ': Incomplete LU decomposition']);
% [L,U] = ilu(Mmod);
disp([datestr(clock) ': Solve for Stream Function']);
% psi   = bicgstab(Mmod,rmod,1e-12,10*nnod,L,U);
psi   = Mmod\rmod;
% remove roundoff errors of bottom and top nodes
psi(topnod)=Qin;
psi(botnod)=0;

% ============== END STREAM FUNCTION CALCULATION ==================================
h=reshape(h,nx(2)+1,nx(1)+1);
psi=reshape(psi,nx(2)+1,nx(1)+1);
contour(X,Y,psi,50,'k');
hold on
contour(X,Y,h,100);
hold off
box on;
daspect([1 1 1]);
xlabel('x_1');ylabel('x_2');
caxis([0 phiin]);
colorbar;
title('flow net')
drawnow
% print -djpeg100 -r300 flownet.jpg

% ============== BEGIN GRID CONSTRUCTION ==========================================
disp([datestr(clock) ': Construction of Streamline-Oriented Grid']);
[net,doagain] = slgrid(ntube,nsec,nx,dx,X,Y,psi,h,phiin,Qin);
end
% ============== END GRID CONSTRUCTION ============================================

% pcolor(net.x,net.y,net.phi);
% set(gca,'dataaspectratio',[1 1 1]);
% shading faceted;
% cb=colorbar;
% ylabel(cb,'h [m]')
% box on;
% daspect([1 1 1]);
% xlabel('x_1');ylabel('x_2');
% title('streamline oriented grid')
% set(gca,'outerposition',[0 1/3+0.025 1 1/3-0.025])
% drawnow;

% ============== BEGIN TRANSPORT PREPARATION ======================================
nnet = ntube*nsec; % number of cells in the net

% streamline oriented grid 2: storage matrix
disp([datestr(clock) ': Evaluate Volume of Streamline-Oriented Elements']);
[area,porarea]=cellarea(net,ntube,nsec,poros);
Mstore = spdiags(porarea,0,nnet,nnet);

% streamline oriented grid 3: mobility matrix
disp([datestr(clock) ': Evaluate Mobility Matrix']);
Mmob = mob_mat(ntube,nsec,Qin,net,al,at,Dp,porarea./area,1);
% ============== END TRANSPORT PREPARATION ========================================

% ============== GROUNDWATER AGE EQUATION =========================================
disp([datestr(clock) ': Temporal-Moment Generation Equations']);
% LU decomposition
% disp([datestr(clock) ': Incomplete LU decomposition']);
% [L,U] = ilu(Mmob);
rhs=porarea(:);
disp([datestr(clock) ': Solve for Mean Groundwater Age']);
% m1=bicgstab(Mmob,rhs,1e-12,10*nnod,L,U);
m1=Mmob\rhs;
m1=reshape(m1,ntube,nsec);

pcolor(net.x,net.y,[m1 m1(:,end);m1(end,:) m1(end,end)]);
shading flat;
set(gca,'dataaspectratio',[1 1 1]);
xlabel('x_1 [m]');ylabel('x_2 [m]');
cb=colorbar;ylabel(cb,'\mu_\tau [s]')
title('mean groundwater age')
drawnow
% print -djpeg100 -r300 gw_age.jpg

% ============== TRANSIENT CONCENTRATIONS =========================================
disp([datestr(clock) ': Solve for Transient Concentration']);

% total time
tend  = 2*max(m1(:));
if CN<1
   % determine time step size so that maximum Courant number is unity
   dt = min(porarea)/Qin*ntube;
else
   % determine time step size so that mean Courant number is unity
   dt = mean(porarea)/Qin*ntube;
end
% Remark: Use Crank-Nicolson for transport
% (Mstore/dt + CN*Mmob)*c_new = (Mstore/dt - (1-CN)*Mmob)*c_old + r_source
Mleft  = Mstore/dt + CN*Mmob;
Mright = Mstore/dt - (1-CN)*Mmob;

% initialization
c = zeros(ntube*nsec,1);

% specific discharge per stream tube
invec = zeros(nnet,1);
invec(1:ntube)=Qin/ntube;

cin = zeros(nnet,1);
cin(1:ntube)=1;

% LU decomposition
% disp([datestr(clock) ': Incomplete LU decomposition']);
% [L,U] = ilu(Mleft);
disp([datestr(clock) ': LU decomposition']);
[L_M,U_M,P_M,Q_M,R_M] = lu(Mleft);

disp([datestr(clock) ': Solve for Concentration']);
tvec=dt:dt:tend;
BTC=ones(ntube,length(tvec));
cplot=reshape(c,ntube,nsec);
if plottransient
    figure(2)
    set(gcf,'outerposition',get(0,'screensize'),'paperunits',...
        'centimeters','paperposition',[0 0 20 10])
    colormap('jet')
    iplot=0;
end
for ii=1:length(tvec)
    t=tvec(ii);
    disp([datestr(clock) ': time =' num2str(t) ' s']);
    days   =floor(t/86400);
    hours  =floor((t-86400*days)/3600);
    minutes=floor((t-86400*days-3600*hours)/60);
    seconds=floor(t-86400*days-3600*hours-60*minutes);
    % old time-point contribution to BTC
    BTCold=cplot(:,end);
    % determine the right-hand side vector
    rhs=Mright*c + ones(nnet,1).*invec.*cin;
    % evaluate the concentration
%     c=bicgstab(Mleft,rhs,1e-12,10*nnod,L,U,c);
    c=Q_M*(U_M\(L_M\(P_M*(R_M\rhs))));
    cplot=reshape(c,ntube,nsec);
    % store the BTC
    BTC(:,ii)=(1-CN)*BTCold+CN*cplot(:,end);
    if min(BTC(:,ii))>0.999,break,end
    if plottransient && mod(ii,40)==1
       pcolor(net.x,net.y,[cplot cplot(:,end);cplot(end,:) cplot(end,end)]);
       caxis([0 1])
       shading flat;
       set(gca,'dataaspectratio',[1 1 1]);
       xlabel('x_1 [m]');ylabel('x_2 [m]');
       cb=colorbar;ylabel(cb,'c/c_{in} [-]')
       title(sprintf('concentration at %3.3id %2.2ih %2.2im %2.2is',[days,hours,minutes,seconds]));
       drawnow;
       iplot=iplot+1;
%        print('-djpeg','-r130',sprintf('picture%4.4i.jpg',iplot))
    end
end
tvec=tvec(1:ii);
BTC=BTC(:,1:ii);

% take derivative to obtain travel-time pdf of each streamtube
ptau=zeros(ntube,length(tvec));
ptaurel=zeros(ntube,length(tvec));
% time scaled by mean groundwater age
trel=1./m1(:,end)*tvec;
for ii=1:ntube
    ptau(ii,:)=gradient(BTC(ii,:),tvec);
    ptaurel(ii,:)=gradient(BTC(ii,:),trel(ii,:));
end
figure(3)
set(gcf,'outerposition',get(0,'screensize'),'paperunits','centimeters',...
    'paperposition',[0 0 20 10])
subplot(1,2,1)
plot(tvec,ptau,'color',[.5 .5 .5])
hold on
plot(tvec,mean(ptau),'color','k','linewidth',2);
hold off
xlabel('\tau [s]')
ylabel('p(\tau) [1/s]')
box on
title('Local Travel-Time Distributions at the Outlet Face')
axis tight
subplot(1,2,2)
plot(trel,ptaurel,'color',[.5 .5 .5])
xlabel('\tau/\mu_\tau [-]')
ylabel('p(\tau/\mu_\tau) [-]')
box on
title('Local Travel-Time Distributions Scaled by Local Mean Groundwater Age')
xlim([0 tvec(end)/mean(m1(:,end))]);
ylim([0,max(ptaurel(:))]);
% print -djpeg100 -r300 BTC.jpg

figure(4)
plot(tvec,mean(ptau),tvec,std(ptau))
xlabel('\tau [s]')
legend('mean p(\tau)','std. of p(\tau)')

% save BTC.mat BTC tvec ptau trel ptaurel

%%%% REPLACEMENT SCENARIO %%% A+B > C with 1:1:1 stoichiometry
%%%% Initial concentration of B: 1
%%%% Injected concentration of A: 1

% Use local breakthrough curves
BTC_A=2*BTC-1; BTC_A(BTC_A<0)=0;
BTC_B=1-2*BTC; BTC_B(BTC_B<0)=0;
BTC_C=BTC-BTC_A;

% use averaged BTC of mixing ratio
BTC_Am=2*mean(BTC)-1; BTC_Am(BTC_Am<0)=0;
BTC_Bm=1-2*mean(BTC); BTC_Bm(BTC_Bm<0)=0;
BTC_Cm=mean(BTC)-BTC_Am;

figure(5)
subplot(2,1,1)
plot(tvec,BTC_A,'color',[.5 .5 1])
hold on
plot(tvec,BTC_B,'color',[1 .5 .5])
plot(tvec,BTC_C,'color',[.5 1 .5])
plot(tvec,mean(BTC_A),'color','b','linewidth',4)
plot(tvec,mean(BTC_B),'color','r','linewidth',4)
plot(tvec,mean(BTC_C),'color','g','linewidth',4)
hold off
xlabel('t [s]')
ylabel('c_{reac}')
box on
title('Reactive-Species Concentrations at the Outlet Face')
axis tight

subplot(2,1,2)
plot(tvec,mean(BTC_A),'color','b','linewidth',2)
hold on
plot(tvec,BTC_Am,'b--','linewidth',2)
plot(tvec,mean(BTC_B),'color','r','linewidth',2)
plot(tvec,BTC_Bm,'r--','linewidth',2)
plot(tvec,mean(BTC_C),'color','g','linewidth',2)
plot(tvec,BTC_Cm,'g--','linewidth',2)
hold off
legend('correct average','from average mixing ratio')
xlabel('t [s]')
ylabel('c_{reac}')
box on
title('Mean Reactive-Species Concentrations at the Outlet Face')
axis tight

